File: Leetcode/Weekly Contest 233/WC233-C.py
class Solution:
    def maxValue(self, n: int, index: int, maxSum: int) -> int:
        def getSum(len, K) -> int:
            if K > len:
                return K * (K+1) // 2 - (K-len) * (K-len+1) // 2
            else:
                return K * (K+1) // 2 + len - K

        ans, l, r = 0, 1, maxSum
        while l <= r:
            K = (l+r) // 2;
            use = getSum(index+1, K) + getSum(n-index, K) - K
            if use <= maxSum:
                ans = K
                l = K + 1
            else:
                r = K - 1
        return ans

File: Leetcode/Weekly Contest 233/WC233-C.cpp
class Solution {
public:
	long long getSum(long long len, long long K)
	{
		if (K > len)
			return (K*(K + 1) / 2ll) - (K - len) * (K - len + 1) / 2ll;
		else
			return K * (K + 1) / 2ll + len - K;
	}

	int maxValue(int n, int index, int maxSum) {
		int ans = 0;
		int l = 1, r = 1e9;
		while (l <= r)
		{
			int k = (l + r) / 2;

			int leftLen = index + 1;
			int rightLen = n - index;

			long long useSum = getSum(leftLen, k) + getSum(rightLen, k) - k;
			if (useSum <= (long long)maxSum)
			{
				ans = k;
				l = k + 1;
			}
			else r = k - 1;
		}

		return ans;
	}
};